File: sprint5/By lessons/lesson10code2/lesson10code2.js
function popMax(heap) {
    const result = heap[0];
    heap[0] = heap[heap.length - 1];
    heap.pop();
    siftDown(heap, 0);
    return result;
}

function siftDown(heap, index) {
    const left = 2 * index + 1;
    const right = 2 * index + 2;

    if (left >= heap.length) {
        return;
    }

    let indexLargest = left;
    if (right < heap.length && heap[left] < heap[right]) {
        indexLargest = right;
    }

    if (heap[index] < heap[indexLargest]) {
        [heap[index], heap[indexLargest]] = [heap[indexLargest], heap[index]];
        siftDown(heap, indexLargest);
    }
}

const heap = [8, 5, 3];
const maxValue = popMax(heap);
console.log(maxValue); // Output: 8
console.log(heap);     // Output: [5, 3]

File: sprint5/By lessons/lesson10code2/lesson10code2.py
def pop_max(heap):
    result = heap[0]
    heap[0] = heap[len(heap) - 1]
    heap.pop()
    sift_down(heap, 0)
    return result

def sift_down(heap, index):
    left = 2 * index + 1
    right = 2 * index + 2

    if len(heap) <= left:
        return

    index_largest = right if right < len(heap) and heap[left] < heap[right] else left

    if heap[index] < heap[index_largest]:
        heap[index], heap[index_largest] = heap[index_largest], heap[index]
        sift_down(heap, index_largest)
